### PR TITLE
React to changes in HTML abstractions

### PR DESCRIPTION
**Pull request by rynowak**
This change implements the new API for flattening content in Razor.

Also, some optimizations to avoid allocations on paths where we need
to encode HTML attribute values.

### CODE REVIEW

File: Razor.sln [hunk 1]
```diff
@@ -43,6 +46,14 @@ Global
...
 		{E3A2A305-634D-4BA6-95DB-AA06D6C442B0}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{E3A2A305-634D-4BA6-95DB-AA06D6C442B0}.Release|Any CPU.Build.0 = Release|Any CPU
+		{68A28E4A-3ADE-4187-9625-4FF185887CB3}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{68A28E4A-3ADE-4187-9625-4FF185887CB3}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{68A28E4A-3ADE-4187-9625-4FF185887CB3}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{68A28E4A-3ADE-4187-9625-4FF185887CB3}.Release|Any CPU.Build.0 = Release|Any CPU
+		{DD2CE416-765E-4000-A03E-C2FF165DA1B6}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{DD2CE416-765E-4000-A03E-C2FF165DA1B6}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{DD2CE416-765E-4000-A03E-C2FF165DA1B6}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{DD2CE416-765E-4000-A03E-C2FF165DA1B6}.Release|Any CPU.Build.0 = Release|Any CPU
```
Comment (rynowak) [inline Razor.sln]: Ignore #YOLO

File: global.json [hunk 2]
```diff
@@ -1,3 +1,3 @@
 {
-    "projects": ["src"]
+    "projects": ["src", "d:\\k\\HtmlAbstractions\\src"]
```
Comment (rynowak) [inline global.json]: MOAR IGNOAR

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs [hunk 3]
```diff
@@ -149,6 +149,29 @@ public override TagHelperContent AppendHtml(IHtmlContent htmlContent)
         }
 
         /// <inheritdoc />
+        public override void CopyTo(IHtmlContentBuilder destination)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: Will add test

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs [hunk 4]
```diff
@@ -129,6 +129,12 @@ public TagHelperContent AppendFormat(IFormatProvider provider, string format, pa
         public abstract TagHelperContent Clear();
 
         /// <summary>
+        /// Copies the content to another <see cref="IHtmlContentBuilder"/>.
+        /// </summary>
+        /// <param name="destination">The destination <see cref="IHtmlContentBuilder"/>.</param>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs]: inheritdoc

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 5]
```diff
@@ -310,21 +450,30 @@ public void WriteTo(TextWriter writer, HtmlEncoder encoder)
...
+
+                                var stringValue = stringWriter.ToString();
+                                writer.Write(stringValue);
+                            }
                         }
                     }
                     else if (value != null)
                     {
-                        encoder.Encode(writer, value.ToString());
+                        writer.Write(value.ToString());
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: This is a copy-paste error, will fix

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 6]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
             return _getChildContentAsync(useCachedResult, encoder);
         }
 
-        /// <inheritdoc />
-        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        void IHtmlContentContainer.CopyTo(IHtmlContentBuilder destination)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Will add tests for new functionality

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs [hunk 7]
```diff
@@ -149,6 +149,29 @@ public override TagHelperContent AppendHtml(IHtmlContent htmlContent)
...
+        {
+            for (var i = 0; i < Buffer.Count; i++)
+            {
+                var entry = Buffer[i];
+                if (entry == null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: aspnet/HttpAbstractions#11 doesn't special case `null`. Should it?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs [hunk 8]
```diff
@@ -149,6 +149,29 @@ public override TagHelperContent AppendHtml(IHtmlContent htmlContent)
...
+                }
+
+                var stringValue = entry as string;
+                if (stringValue != null)
+                {
+                    destination.Append(stringValue);
+                }
+                else
+                {
+                    destination.AppendHtml((IHtmlContent)entry);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: Should `IHtmlContentContainer` be special-cased here?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs [hunk 9]
```diff
@@ -167,5 +173,17 @@ IHtmlContentBuilder IHtmlContentBuilder.Clear()
         {
             return Clear();
         }
+
+        /// <inheritdoc />
+        void IHtmlContentContainer.MoveTo(IHtmlContentBuilder destination)
+        {
+            if (destination == null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs]: Good point. `null` checks are missing in `CopyTo()` as well as aspnet/HttpAbstractions#11 in general.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 10]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+                            destination.AppendHtml(htmlEncodedString);
+                        }
+                        else
+                        {
+                            destination.AppendHtml(new AttributeContent(htmlContent));
+                        }
+                    }
+                    else if (value != null)
+                    {
+                        destination.Append(value.ToString());
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Why `ToString()`? IOW what can `value` be other than a `string` or an `IHtmlContent` implementation?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 11]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+                    {
+                        continue;
+                    }
+
+                    destination.AppendHtml("=\"");
+                    var value = attribute.Value;
+                    var htmlContent = value as IHtmlContent;
+                    if (htmlContent != null)
+                    {
+                        // Perf: static text in a bound attribute go down this path. Avoid allocating if possible (common case).
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Note we'll only go down this path if the _entire_ attribute value is one bit of static text. May need to make quote escaping more first-class to handle static text in mixed cases.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 12]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+                destination.AppendHtml(">");
+            }
+
+            _postElement?.CopyTo(destination);
+        }
+
+        void IHtmlContentContainer.MoveTo(IHtmlContentBuilder destination)
         {
-            if (writer == null)
+            ((IHtmlContentBuilder)_preContent)?.MoveTo(destination);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: `_preElement`

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 13]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+            {
+                destination.AppendHtml("</");
+                destination.AppendHtml(TagName);
+                destination.AppendHtml(">");
+            }
+
+            _postElement?.CopyTo(destination);
+        }
+
+        void IHtmlContentContainer.MoveTo(IHtmlContentBuilder destination)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Lots of duplication. Why not call `CopyTo()` then `Clear()` the non-`null` fields?

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 14]
```diff
@@ -981,7 +981,9 @@ public void WriteTo_WritesFormattedTagHelper_HtmlTextWriter(TagHelperOutput outp
             // Arrange
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Update test name since you're no longer testing handling of `HtmlTextWriter`.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 15]
```diff
@@ -356,5 +505,30 @@ public void WriteTo(TextWriter writer, HtmlEncoder encoder)
...
+            }
+
+            public IHtmlContent Inner { get; }
+
+            public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+            {
+                // There's no way of tracking the attribute value quotations in the Razor source. Therefore, we
+                // must escape any IHtmlContent double quote values in the case that a user wrote:
+                // <p name='A " is valid in single quotes'></p>
+                using (var stringWriter = new StringWriter())
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Suggest allocating the `StringWriter()` in the containing class (lazily), passing into the constructor, and clearing on every `WriteTo()` call. That should somewhat reduce the explosion of `char[]`, `StringBuilder` and `StringWriter` instances.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 16]
```diff
@@ -356,5 +505,30 @@ public void WriteTo(TextWriter writer, HtmlEncoder encoder)
...
         }
+
+        private class AttributeContent : IHtmlContent
+        {
+            public AttributeContent(IHtmlContent inner)
+            {
+                Inner = inner;
+            }
+
+            public IHtmlContent Inner { get; }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Surprised this isn't a `private` field. Where is this needed / used?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 10 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Never mind. Forgot for a sec how general `TagHelperAttribute.Value` is.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 17]
```diff
@@ -276,14 +276,154 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+                            destination.AppendHtml(htmlEncodedString);
+                        }
+                        else
+                        {
+                            destination.AppendHtml(new AttributeContent(htmlContent));
+                        }
+                    }
+                    else if (value != null)
+                    {
+                        destination.Append(value.ToString());
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Slightly tangential to this PR: Some `ToString()` implementations are quite expensive. Bet we'd see real savings in any scenario involving `DateTimeOffset` or similar `value`s if `TagHelperAttribute` exposed `string ValueString { get; }` but left it `null` when `Value is IHtmlContent`.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 13 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Discussed offline about need to recursively move everything down the tree.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: `HtmlContentBuilder` doesn't ever contain a null. The taghelper things do. I dunno why, but it's not part of this change.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: Yes. Added

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperContent.cs]: Those APIs never store `null` so they don't need it.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 11 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: This proved to be an extremely profitable optimization in the case I was looking at.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Fixed

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 15 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Yeah, that actually seems really nice.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :+1: awesome suggestion. Works out really well.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 16 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: I just code things that way by default. I'll change it #YOLO

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 17 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Yeah I'd love to dig into this more. I well like the way attributes are handled in general is slightly suboptimal.

**Comment by rynowak** [comment]
@dougbu updated

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs [hunk 18]
```diff
@@ -149,6 +149,84 @@ public override TagHelperContent AppendHtml(IHtmlContent htmlContent)
...
+                {
+                    entryAsContainer.MoveTo(destination);
+                }
+                else
+                {
+                    destination.AppendHtml((IHtmlContent)entry);
+                }
+            }
+
+            Buffer.Clear();
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: `Clear()` instead of `Buffer.Clear()`?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: Why?

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: It doesn't matter. I just thought it is better to do it all in one place.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs]: The reason why I did `Buffer.Clear` is that this method is already operating directly on the buffer. It would mix levels of abstraction to call `Clear()`

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 19]
```diff
@@ -276,14 +276,87 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
             return _getChildContentAsync(useCachedResult, encoder);
         }
 
-        /// <inheritdoc />
-        public void WriteTo(TextWriter writer, HtmlEncoder encoder)
+        void IHtmlContentContainer.CopyTo(IHtmlContentBuilder destination)
         {
-            if (writer == null)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Add `null` check for `destination`.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 20]
```diff
@@ -276,14 +276,87 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+                destination.AppendHtml(TagName);
+                destination.AppendHtml(">");
+            }
+
+            _postElement?.CopyTo(destination);
+        }
+
+        void IHtmlContentContainer.MoveTo(IHtmlContentBuilder destination)
+        {
+            _preElement?.MoveTo(destination);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Add `null` check for `destination`.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 21]
```diff
@@ -276,14 +276,87 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(TagName);
+
+            if (!isTagNameNullOrWhitespace)
+            {
+                destination.AppendHtml("<");
+                destination.AppendHtml(TagName);
+
+                CopyAttributesTo(destination);
+                _attributes?.Clear();
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Why call `Clear()` in the `CopyTo()` method?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 22]
```diff
@@ -276,14 +276,87 @@ public Task<TagHelperContent> GetChildContentAsync(bool useCachedResult, HtmlEnc
...
+
+            var isTagNameNullOrWhitespace = string.IsNullOrWhiteSpace(TagName);
+
+            if (!isTagNameNullOrWhitespace)
+            {
+                destination.AppendHtml("<");
+                destination.AppendHtml(TagName);
+                
+                CopyAttributesTo(destination);
+
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: `Clear()` the `_attributes`.

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs [hunk 23]
```diff
@@ -356,5 +438,76 @@ public void WriteTo(TextWriter writer, HtmlEncoder encoder)
...
+                destination.AppendHtml(attribute.Name);
+
+                if (attribute.Minimized)
+                {
+                    continue;
+                }
+
+                destination.AppendHtml("=\"");
+                var value = attribute.Value;
+                var htmlContent = value as IHtmlContent;
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Is it impossible for `_attributes` to contain an `IHtmlContentBuilder`. Unsure this is the Right Thing:tm: in the `MoveTo()` case if not.

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 24]
```diff
@@ -972,16 +972,15 @@ public void WriteTo_WritesFormattedTagHelper(TagHelperOutput output, string expe
             Assert.Equal(expected, writer.ToString(), StringComparer.Ordinal);
         }
 
-        // This tests a separate code path that's used by THO when the writer is an HtmlTextWriter.
-        // The output should be the same, but we do some specific perf optimizations on this path.
         [Theory]
         [MemberData(nameof(WriteTagHelper_InputData))]
-        public void WriteTo_WritesFormattedTagHelper_HtmlTextWriter(TagHelperOutput output, string expected)
+        public void CopyTo_CopiesToBuilder(TagHelperOutput output, string expected)
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Add either a test that you can call `CopyTo()` twice and get the same result or a test that confirms `Attributes.Count` before equals `Attributes.Count` after.

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs [hunk 25]
```diff
@@ -994,10 +993,48 @@ public void WriteTo_WritesFormattedTagHelper_HtmlTextWriter(TagHelperOutput outp
...
             // Assert
-            Assert.Equal(expected, inner.ToString(), StringComparer.Ordinal);
+            buffer.WriteTo(writer, testEncoder);
+
+            Assert.True(output.PreElement.IsEmpty);
+            Assert.True(output.PreContent.IsEmpty);
+            Assert.True(output.Content.IsEmpty);
+            Assert.True(output.PostContent.IsEmpty);
+            Assert.True(output.PostElement.IsEmpty);
+            Assert.Empty(output.Attributes);
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Odd. How does this assertion succeed?

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 13 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :+1:

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 14 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Thx

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 19 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :+1:

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 20 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :+1:

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 21 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: Oops!

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 23 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: I have no idea really. I'll ask whatshisface when he shows up

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 24 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: Added assertion to this test

**Comment by rynowak** [comment]
:updated: - need to discuss https://github.com/aspnet/Razor/pull/694#discussion_r54317974

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 23 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: The answer from @NTaylorMullen is that this will generally be an `HtmlEncodedString` or string + needs encoding.

The user could do something weird here like put a view buffer inside of an attribute. I think we're we're doing here is fine, we'd be suboptimal, but not incorrect if you did that.

**Comment by rynowak** [comment]
Updated again - and fixed test

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 25 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: It does now that I fixed the implementation

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 22 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: thx

File: src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs (hunk 23 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/TagHelperOutput.cs]: :ok_hand:

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 24 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: thx

File: test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs (hunk 25 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Razor.Runtime.Test/TagHelpers/TagHelperOutputTest.cs]: :smiley_cat:

**Comment by dougbu** [comment]
:shipit:

**Comment by rynowak** [comment]
will check in once MVC part is signed off.
